// snarkyjs/dist/web/snarky/gen/parties.js
// @generated this file is auto-generated - don't edit it directly
import { convertStringWithHashToJson, convertEventsToJson, convertStringWithHashToFields, convertEventsToFields, } from '../parties-leaves';
import { toJson, toFields } from '../parties-helpers';
import * as Json from './parties-json';
import { jsLayout } from './js-layout';
export { Parties, Party };
export { Json };
export * from '../parties-leaves';
let jsonConverters = {
    StringWithHash: convertStringWithHashToJson,
    Events: convertEventsToJson,
};
let fieldsConverters = {
    StringWithHash: convertStringWithHashToFields,
    Events: convertEventsToFields,
};
let Parties = {
    toJson(parties) {
        return toJson(jsLayout.Parties, parties, jsonConverters);
    },
    toFields(parties) {
        return toFields(jsLayout.Parties, parties, fieldsConverters);
    },
};
let Party = {
    toJson(party) {
        return toJson(jsLayout.Party, party, jsonConverters);
    },
    toFields(party) {
        return toFields(jsLayout.Party, party, fieldsConverters);
    },
};
