// eslint-plugin-snarkyjs/build/tests/rules/no-if-in-circuit.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rule_tester_1 = require("../rule-tester");
const no_if_in_circuit_1 = __importDefault(require("../../src/rules/no-if-in-circuit"));
const message = 'noIfInCircuit';
rule_tester_1.ruleTester.run('no-if-in-circuit', no_if_in_circuit_1.default, {
    valid: [
        {
            code: `
      class Foo {
        async bar() {
          if(true) {};
        }
      }`,
        },
        {
            code: `
      function testIf() { if (true) {}; };
      class Foo {
        async bar() {
          testIf();
        }
      }`,
        },
        {
            code: `
      class Foo {
        @method async bar() {}
      }`,
        },
    ],
    invalid: [
        {
            code: `
      class Foo {
     	  @method async bar() {
          if (true) {};
        }
      }`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      function testIf() { if (true) {}; };
      class Foo {
        @method async bar() {
         testIf();
        }
      }`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      let testIf = () => { if (true); };
      function indirectIf() { testIf(); }
      class Foo {
        @method async myMethod() {
          indirectIf();
        }
      }
      `,
            errors: [{ messageId: message }],
        },
    ],
});
