// eslint-plugin-snarkyjs/build/tests/rules/no-throw-in-circuit.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rule_tester_1 = require("../rule-tester");
const no_throw_in_circuit_1 = __importDefault(require("../../src/rules/no-throw-in-circuit"));
const message = 'noThrowInCircuit';
rule_tester_1.ruleTester.run('no-throw-in-circuit', no_throw_in_circuit_1.default, {
    valid: [
        {
            code: `
      class Foo {
        async bar() {
          throw "foobar";
        }
      }`,
        },
        {
            code: `
      function testThrow() { throw "test" };
      class Foo {
        async bar() {
          testThrow();
        }
      }`,
        },
        {
            code: `
      class Foo {
        @method async bar() {}
      }`,
        },
    ],
    invalid: [
        {
            code: `
      class Foo {
        @method async bar() {
          throw "foobar";
        }
      }`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      function testThrow() { throw "test" };
      class Foo {
        @method async bar() {
          testThrow();
        }
      }`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      let testThrow = () => { throw "test"; };
      function indirectThrow() { testThrow(); }
      class Foo {
        @method async myMethod() {
          indirectThrow();
        }
      }
      `,
            errors: [{ messageId: message }],
        },
    ],
});
