// eslint-plugin-snarkyjs/build/tests/rules/no-ternary-in-circuit.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rule_tester_1 = require("../rule-tester");
const no_ternary_in_circuit_1 = __importDefault(require("../../src/rules/no-ternary-in-circuit"));
const message = 'noTernaryInCircuit';
rule_tester_1.ruleTester.run('no-ternary-in-circuit', no_ternary_in_circuit_1.default, {
    valid: [
        {
            code: `
      class Foo {
        async bar() {
          let a = true ? true : false;
        }
      }`,
        },
        {
            code: `
      function testTernary() { let a = true ? true : false; };
      class Foo {
        async bar() {
          testTernary();
        }
      }`,
        },
        {
            code: `
      class Foo {
        @method async bar() {}
      }`,
        },
    ],
    invalid: [
        {
            code: `
      class Foo {
        @method async bar() {
          let a = true ? true : false;
        }
    	}`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      function testTernary() { let a = true ? true : false; };
      class Foo {
        @method async bar() {
          testTernary();
        }
    	}`,
            errors: [{ messageId: message }],
        },
        {
            code: `
      let testTernary = () => { true ? true : false; };
      function indirectTernary() { testTernary(); }
      class Foo {
        @method async myMethod() {
          indirectTernary();
        }
      }
      `,
            errors: [{ messageId: message }],
        },
    ],
});
